File: PRACTICAL_05_Transformation_CS_src/Quaternion.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace cube
{
    class Quaternion
    {
        private double x, y, z, w;

        public Quaternion()
        {
            w = 0; x = 0; y = 0; z = 0;
        }
        public Quaternion(double _w, double _x, double _y, double _z)
        {
            w = _w; x = _x; y = _y; z = _z;
        }

        public Quaternion(double _w, Vector3 v)
        {
            w = _w; x = v.X; y = v.Y; z = v.Z;
        }

        public override string ToString()
        {
            return "("+ w.ToString("g3") + "," + x.ToString("g3") + "i," + y.ToString("g3") + "j," + z.ToString("g3") + "k)";
        }

        public Double Modulus
        { get{ return Math.Sqrt(w*w+x*x+y*y+z*z); }
        }
        public Quaternion Normalise()
        {
            double m = w * w + x * x + y * y + z * z;
            if (m > 0.001)
            {
                m = Math.Sqrt(m);
                return new Quaternion(w / m, x / m, y / m, z / m);
            }
            else
            {
                return new Quaternion(1, 0, 0, 0);
            }
        }

        public Quaternion Conjugate()
        {
           return new Quaternion  (w,-x,-y,-z);
        }

        public void FromAxisAngle(Vector3 axis, double angleRadian)
        {
            double m = axis.Length();
            if (m > 0.0001)
            {
                double ca = Math.Cos(angleRadian / 2);
                double sa = Math.Sin(angleRadian / 2);
                x = axis.X / m * sa;
                y = axis.Y / m * sa;
                z = axis.Z / m * sa;
                w = ca;
            }
            else
            {
                w = 1; x = 0; y = 0; z = 0;
            }
        }

        public Quaternion Multiply(Quaternion _q)
        {
            double nw = _q.w *  w - _q.x *  x - _q.y *  y - _q.z *  z;
            double nx = _q.w *  x + _q.x *  w + _q.y *  z - _q.z *  y;
            double ny = _q.w *  y + _q.y *  w + _q.z *  x - _q.x *  z;
            double nz = _q.w *  z + _q.z *  w + _q.x *  y - _q.y *  x;
            return new Quaternion(nw, nx, ny, nz);
        }

        public Quaternion Copy()
        {
            return new Quaternion(x, y, z, w);
        }
        //                  -1
        // V'=q*V*q     ,
    
        public Vector3 Rotate(Vector3 pt, int _angle)
        {
            Quaternion axis, rotate;
            axis = this.Normalise();
            double angleRad = Math.PI / 180 * _angle;
            rotate = new Quaternion(Math.Cos(angleRad / 2), Math.Sin(angleRad / 2) * axis.x, Math.Sin(angleRad / 2) * axis.y, Math.Sin(angleRad / 2) * axis.z);
            Quaternion conjugate = rotate.Conjugate();
            Quaternion qNode = new Quaternion(0, pt.X, pt.Y, pt.Z);
            qNode = rotate * qNode * conjugate;
            return new Vector3(qNode.x, qNode.y, qNode.z);        
        }

        // Multiplying q1 with q2 is meaning of doing q2 firstly then q1
        public static Quaternion operator +(Quaternion q1, Quaternion q2)
        {
            return new Quaternion(q1.w + q2.w, q1.x + q2.x, q1.y + q2.y, q1.z + q2.z);
        }

        public static Quaternion operator -(Quaternion q1)
        {
            return new Quaternion(-q1.w, -q1.x,-q1.y, -q1.z);
        }

        public static Quaternion operator -(Quaternion q1, Quaternion q2)
        {
            return new Quaternion(q1.w - q2.w, q1.x - q2.x, q1.y - q2.y, q1.z - q2.z);
        }
#region multiply
        
        public static Quaternion operator *(Quaternion q1, double s)
        {
            return new Quaternion(q1.w * s, q1.x * s, q1.y * s, q1.z * s);
        }
        public static Quaternion operator *(Quaternion q1, float s)
        {
            return new Quaternion(q1.w * s, q1.x * s, q1.y * s, q1.z * s);
        }
        public static Quaternion operator *(Quaternion q1, int s)
        {
            return new Quaternion(q1.w * s, q1.x * s, q1.y * s, q1.z * s);
        }
        public static Quaternion operator *(double s,Quaternion q1)
        {
            return new Quaternion(q1.w * s, q1.x * s, q1.y * s, q1.z * s);
        }
        public static Quaternion operator *(float s, Quaternion q1)
        {
            return new Quaternion(q1.w * s, q1.x * s, q1.y * s, q1.z * s);
        }
        public static Quaternion operator *(int s, Quaternion q1)
        {
            return new Quaternion(q1.w * s, q1.x * s, q1.y * s, q1.z * s);
        }
#endregion       

        public static Quaternion operator *(Quaternion q1, Quaternion q2)
        {
            double nw = q1.w * q2.w - q1.x * q2.x - q1.y * q2.y - q1.z * q2.z;
            double nx = q1.w * q2.x + q1.x * q2.w + q1.y * q2.z - q1.z * q2.y;
            double ny = q1.w * q2.y + q1.y * q2.w + q1.z * q2.x - q1.x * q2.z;
            double nz = q1.w * q2.z + q1.z * q2.w + q1.x * q2.y - q1.y * q2.x;
            return new Quaternion(nw, nx, ny, nz);
        }
    }    
}

File: PRACTICAL_05_Transformation_CS_src/src/Quaternion.cpp
#include "Quaternion.h"


/// <summary>
/// Default contructor
/// </summary>
Quaternion::Quaternion():
	m_w(0),
	m_x(0),
	m_y(0),
	m_z(0)
{
}
/// <summary>
/// constructor that takes in 4 doubles
/// </summary>
/// <param name="w">w component</param>
/// <param name="x">x component</param>
/// <param name="y">y component</param>
/// <param name="z">z component</param>
Quaternion::Quaternion(double w, double x, double y, double z) :
	m_w(w),
	m_x(x),
	m_y(y),
	m_z(z)
{
}
/// <summary>
/// Constructor that takes in the w component and the x,y,z as a vector3
/// </summary>
/// <param name="w">double w component</param>
/// <param name="v">vector3 x,y,z components</param>
Quaternion::Quaternion(double w, Vector3 v) :
	m_w(w),
	m_x(v.getX()),
	m_y(v.getY()),
	m_z(v.getZ())
{
}
/// <summary>
/// Returns a quaternion as string
/// </summary>
/// <returns>string of quaternion</returns>
std::string Quaternion::toString()
{
	return std::string("(" + std::to_string(m_w) + "," + std::to_string(m_x) + "i," + std::to_string(m_y) + "j," + std::to_string(m_z) + "k)");
}
/// <summary>
/// Gets the mod of the quaternion
/// </summary>
/// <returns>modulus</returns>
double Quaternion::modulus() const
{
	return sqrt(m_w * m_w + m_x * m_x + m_y * m_y + m_z * m_z);
}
/// <summary>
/// Mod squared to prevent square rooting for efficiency
/// </summary>
/// <returns>double mod squared</returns>
double Quaternion::modulusSquared() const
{
	return m_w * m_w + m_x * m_x + m_y * m_y + m_z * m_z;
}
/// <summary>
/// normalises the quaternion
/// </summary>
/// <returns>normalised quaternion</returns>
Quaternion Quaternion::normalise() const
{
	double m = modulusSquared();
	if (m > 0.001)
	{
		m = sqrt(m);
		return Quaternion(m_w / m, m_x / m, m_y / m, m_z / m);
	}
	else
	{
		return Quaternion(1, 0, 0, 0);
	}
}
/// <summary>
/// Gets the conjugate of a quaternion
/// </summary>
/// <returns>quaternions conjugate</returns>
Quaternion Quaternion::conjugate() const
{
	return Quaternion(-m_w, -m_x, -m_y, -m_z);
}
/// <summary>
/// make an axis angle as quaternion
/// </summary>
/// <param name="axis">the axis</param>
/// <param name="angleRad">angle in radians</param>
void Quaternion::fromAxisAngle(Vector3 axis, double angleRad)
{
	double m = axis.length();
	if (m > 0.0001)
	{
		double ca = cos(angleRad / 2);
		double sa = sin(angleRad / 2);
		m_x = axis.getX() / m * sa;
		m_y = axis.getY() / m * sa;
		m_z = axis.getZ() / m * sa;
		m_w = ca;
	}
	else
	{
		m_w = 1;
		m_x = 0;
		m_y = 0;
		m_z = 0;
	}
}
/// <summary>
/// Multiply quaternion by another quaternion
/// </summary>
/// <param name="q">the quaternion to multiply by</param>
/// <returns></returns>
Quaternion Quaternion::multiply(const Quaternion& q) const
{
	return Quaternion(  q.m_w *  m_w - q.m_x *  m_x - q.m_y *  m_y - q.m_z *  m_z,
						q.m_w *  m_x + q.m_x *  m_w + q.m_y *  m_z - q.m_z *  m_y,
						q.m_w *  m_y + q.m_y *  m_w + q.m_z *  m_x - q.m_x *  m_z,
						q.m_w *  m_z + q.m_z *  m_w + q.m_x *  m_y - q.m_y *  m_x);
}
/// <summary>
/// Return a rotation vector
/// </summary>
/// <param name="pt">the axis by which to rotate</param>
/// <param name="angle">angle at which to rotate</param>
/// <returns>rotation vector</returns>
Vector3 Quaternion::rotate(const Vector3 &pt, const int &angle) const
{
	Quaternion axis;
	Quaternion rotateQ;
	axis = normalise();
	double angleRad = acos(-1) / 180 * angle;
	rotateQ = Quaternion(cos(angleRad / 2), sin(angleRad / 2) * axis.m_x, sin(angleRad / 2) * axis.m_y, sin(angleRad / 2) * axis.m_z);
	Quaternion conjugate = rotateQ.conjugate();
	Quaternion qNode = Quaternion(0, pt.getX(), pt.getY(), pt.getZ());
	qNode = rotateQ * qNode * conjugate;
	return Vector3(qNode.m_x,qNode.m_y,qNode.m_z);
}
/// <summary>
/// Multiplying one quaternion by another
/// </summary>
/// <param name="q">quaternion to multiply</param>
/// <returns>result of the product of 2 quaternions</returns>
Quaternion Quaternion::operator*(const Quaternion& q) const
{
	return Quaternion(q.multiply(*this));
}
/// <summary>
/// Adding 2 quaternions
/// </summary>
/// <param name="q">quaternion to add</param>
/// <returns>result of the sum of two quaternions</returns>
Quaternion Quaternion::operator+(const Quaternion &q) const
{
	return Quaternion(m_w + q.m_w, m_x + q.m_x, m_y + q.m_y, m_z + q.m_z);
}
/// <summary>
/// Result of subtracting a quaternion
/// </summary>
/// <param name="q">the quaternion to subtract</param>
/// <returns>subtracted quaternion</returns>
Quaternion Quaternion::operator-(const Quaternion &q) const
{
	return Quaternion(m_w - q.m_w, m_x - q.m_x, m_y - q.m_y, m_z - q.m_z);
}
/// <summary>
/// Negates the quaternion
/// </summary>
/// <returns>negated quaternion</returns>
Quaternion Quaternion::operator-() const
{
	return Quaternion(-m_w, -m_x, -m_y, -m_z);
}

